### DISCUSSION TITLE
개발 소요

### DISCUSSION THREAD

**Discussion by RimWorldKorea** [RimWorldKorea/RMK#89] [일반]
@csh1668 

혹시 이런 프로그램 제작이 가능할까요?
깃헙과 무관하게 윈도우 상에서 돌아가는 프로그램입니다

<hr>

![제목 없음1](https://github.com/RimWorldKorea/RMK/assets/150018627/7388243e-0382-48cf-ae02-e22d16b15ffb)

예를 들어서 저런 식으로 폴더 구조가 되어있는데
어떤 최상위 폴더를 투입하면
그 폴더의 모든 하위 폴더에 들어있는 xml 파일들의 element와, 그 xml 파일이 들어있는 폴더의 이름을 이용해서 목록을 만드는 겁니다

![제목 없음2](https://github.com/RimWorldKorea/RMK/assets/150018627/d8182cd0-925a-491c-b643-6b80b7b556ae)

<hr>

<img width="872" alt="화면 캡처 2023-11-16 190157" src="https://github.com/RimWorldKorea/RMK/assets/150018627/250fb55d-31c3-4fa2-be89-91f05d1ef607">

위 이미지를 보면 TorannMagic.MagicRecipeDef 폴더 아래에 xml 파일이 있습니다
각 폴더에 들어있는 xml 파일은 하나일 수도, 여러 개일 수도 있습니다
이런 것들을 가공해서 최종적으로 아래와 같은 형태로 만들어야 합니다

![화면 캡처 2023-11-16 190933](https://github.com/RimWorldKorea/RMK/assets/150018627/6f047a8d-2b49-4a43-a99b-f66260ec6aff)

현재 작업량의 거의 50% 정도를 잡아먹는 과정이라
이걸 한 번에 할 수 있다면 업데이트 속도가 어마어마하게 빨라질 수 있을 것 같아요

**Comment by csh1668** [comment]
나름 간단하게 가능할 거 같은데 다만 궁금한 점이 있어요

최종적으로는 csv와 같은 형태로 만드는 것 같은데, 그 이후의 과정은 어떻게 되는 건가요?
쓰이는 목적을 알면 더욱 목적에 부합하게 만들 수 있을 것 같습니다.

**Reply by RimWorldKorea** [reply]
[알파 추출기](https://drive.google.com/file/d/1g38FuI9qfs-c8OERZC6FGPT7OFtKdU95/view?usp=drive_link)를 사용하면 특정 모드의 번역 데이터를 스프레드시트로 추출이 가능한데,
이렇게 뽑은 시트를 그대로 xml 파일로 변환하는 기능도 있습니다.

위에서 말한 "(폴더 이름)+(요소 이름)" 텍스트가 이 스프레드시트에서 특정 번역문을 관리하는 고유값으로 사용되기 때문에
저런 형태로 만들 수 있다면 시트 형태로 배포되지 않은 번역 파일을 스프레드시트에 손쉽게 매핑할 수 있게 됩니다

[이 파일](https://github.com/RimWorldKorea/RMK/files/13377137/2329011599.xlsx)을 참고하면 도움이 될 것 같습니다.
'Merge_RKTM' 시트와 'Merge_RimMod' 시트에서 각 번역 데이터를 취합한 뒤
'Sheet' 시트에서 통합하는 과정이 담겨있습니다.

**Reply by csh1668** [reply]
@RimWorldKorea 
[net7.0.zip](https://github.com/RimWorldKorea/RMK/files/13378668/net7.0.zip)

급한대로 베타버전 만들어봤습니다.

사용방법:
![스크린샷 2023-11-16 225954](https://github.com/RimWorldKorea/RMK/assets/18442452/490a3e78-b7f6-46ce-9506-2d5c1244981a)
압축 풀고, 파일 탐색기에서 빈공간 쉬프트 우클릭 후 '터미널에서 열기'를 클릭합니다.

`.\RKMTools.exe -xmllister -s "폴더 경로"`
위와 같이 입력합니다
예시)
`.\RKMTools.exe -xmllister -s "C:\Games\Steam\steamapps\common\RimWorld\Mods\RMK-forked\Data\(NWN) Real Fog of War - 2560931731\Languages"`
무조건 Languages 폴더를 지정할 필요는 없습니다. 'Korean (한국어)' 또는 'Korean' 폴더나 그의 상위 폴더면 모두 가능합니다.
`C:\Games\Steam\steamapps\common\RimWorld\Mods\RMK-forked\Data`
이런식으로 전체 지정도 가능하다는 뜻입니다.
전체 지정으로 돌려보니 xml 파일 내부가 잘못되어 있는 것들이 좀 있네요

실행 후 같은 폴더에 result.csv라 이름지어진 파일이 추가됩니다.

만약, 결과물을 다른 곳에 저장하길 원한다면
`.\RKMTools.exe -xmllister -s "목표 폴더" "다른 폴더"`
위와 같이 입력합니다.

만약 실행이 안된다면 아마도? 닷넷 런타임이 안깔려있어서 그런 걸 수도 있습니다.
https://dotnet.microsoft.com/ko-kr/download/dotnet/7.0
여기서 .NET 런타임 7.0.14를 설치해야 할지도 모릅니다.

![스크린샷 2023-11-16 230253](https://github.com/RimWorldKorea/RMK/assets/18442452/4f874edb-db34-4bbd-b459-c83229536022)

피드백 부탁드립니다!

**Reply by RimWorldKorea** [reply]
제가 생각하던 그대로네요!

다만 csv 파일의 텍스트 인코딩 설정을 지정할 수 있는 옵션이 있었으면 좋겠습니다

저는 마이크로소프트 엑셀을 주로 쓰는데
csv 디코딩 디폴트가 ANSI라서 출력 파일을 그대로 쓰면 파일이 전부 깨지더라구요

쓰는 사람마다 필요한 설정이 다를 수 있으니
컨픽 파일 같은데 저장시켜 두고 매 번 입력할 필요가 없게 하는게 어떨까 해요

**Reply by csh1668** [reply]
@RimWorldKorea 
인코딩 문제가 있었네요.. 옛날에도 csv로 하다가 그런 문제가 있었는데 저는 한셀이 기본으로 되어있어서 까먹었었네요 ㅋㅋ

[net7.0.zip](https://github.com/RimWorldKorea/RMK/files/13390191/net7.0.zip)
컨피그 파일 만들었습니다. 
앞으로는 `RMKTools.settings.xml` 파일을 열어서 림월드 xml 수정하듯 설정하면 되고, `RKMTools.exe`로 실행합니다.
주석이 다 달려있으니 아마 큰 어려움이 없을 겁니다.

**Reply by RimWorldKorea** [reply]
지금 써보고 있는 중인데
작업 폴더 경로를 설정 파일에서만 바꿀 수 있게 바뀐 건가요?

<img width="636" alt="화면 캡처 2023-11-18 210825" src="https://github.com/RimWorldKorea/RMK/assets/150018627/030b37e0-787f-4126-8c37-cd8ff069bbc8">

파워쉘(터미널) 안쓰고 exe 파일을 직접 실행시켜도 되던데
이 콘솔창 위에서 모든 작업이 가능하면 더 좋을 것 같아요

    예를 들어서 작업 프로세스를 이런 형태로 구현하면 좋을 것 같아요
        1) 초기 화면에서 1 입력하여 XMLister 진입
        2) 현재 설정 및 지정된 경로 화면에 출력
        3) -path "경로" 입력으로 경로 변경
        4) 변경된 현재 경로 화면에 출력
        5) -dset_path none 또는 -dest_path "경로" 입력으로 출력 경로 변경
        6) 변경된 현재 설정 화면에 출력
        7) -run 입력으로 lister 작동
        8) 다음 입력 대기 또는 -esc로 초기화면 복귀

**Comment by RimWorldKorea** [comment]
의도하지 않은 동작

[Slave Outfits [1.4] - 1522429439.zip](https://github.com/RimWorldKorea/RMK/files/13401795/Slave.Outfits.1.4.-.1522429439.zip)

해당 파일을 XMLister로 돌렸을 때

<img width="713" alt="화면 캡처 2023-11-19 092425" src="https://github.com/RimWorldKorea/RMK/assets/150018627/753a7659-5990-4749-b7df-056431416c75">

위 짤처럼 녹색 부분은 제대로 뽑히는데
당근 부분은 행렬이 바사삭 나뉘어 버리네요

**Comment by RimWorldKorea** [comment]
왜 그런진 모르겠는데
xml 선언문에 UTF-8이라 지정해놓고 정작 파일 인코딩이 UTF-16인게 있더라구요

이런건 처리를 못하고 오류가 터지던데
2번 기능으로 일괄 재인코딩 기능 같은걸 넣어봐도 괜찮을 것 같아요

이런거 해주는 프로그램은 찾으려면 있을 법 하기도 하고
진짜 희귀 케이스라서 지금은 그냥 사례 보고 정도로..

**Comment by RimWorldKorea** [comment]
[MoreFloorsTerrainsCarpet.zip](https://github.com/RimWorldKorea/RMK/files/13401880/MoreFloorsTerrainsCarpet.zip)

해당 파일을 XMLister로 돌렸을 때 주석까지 전부 읽어들여버려요

<img width="1735" alt="화면 캡처 2023-11-19 104315" src="https://github.com/RimWorldKorea/RMK/assets/150018627/8f28b622-a9df-4979-a83b-e95277074cfb">

**Comment by csh1668** [comment]
[net7.0.zip](https://github.com/RimWorldKorea/RMK/files/13403021/net7.0.zip)

- 요청하신대로 이제 콘솔창 작동 로직을 개선했습니다.

> 의도하지 않은 동작
> 
> [Slave Outfits [1.4] - 1522429439.zip](https://github.com/RimWorldKorea/RMK/files/13401795/Slave.Outfits.1.4.-.1522429439.zip)
> 
> 해당 파일을 XMLister로 돌렸을 때
> 
> <img alt="화면 캡처 2023-11-19 092425" width="713" src="https://user-images.githubusercontent.com/150018627/284029530-753a7659-5990-4749-b7df-056431416c75.png">
> 위 짤처럼 녹색 부분은 제대로 뽑히는데 당근 부분은 행렬이 바사삭 나뉘어 버리네요
 - 이 부분은 csv의 요소가 큰 따옴표 안에 있는데 ("123", "321") 그 안에 큰 따음표가 있어서 발생하는 문제로 보이네요.
구분자를 바꿀 수 있도록 설정을 따로 빼둠으로써 해결했습니다. 기본값은 ','인데 '|'이런걸로 바꾸고 하면 될 것 같아요.



> 왜 그런진 모르겠는데 xml 선언문에 UTF-8이라 지정해놓고 정작 파일 인코딩이 UTF-16인게 있더라구요
> 
> 이런건 처리를 못하고 오류가 터지던데 2번 기능으로 일괄 재인코딩 기능 같은걸 넣어봐도 괜찮을 것 같아요
> 
> 이런거 해주는 프로그램은 찾으려면 있을 법 하기도 하고 진짜 희귀 케이스라서 지금은 그냥 사례 보고 정도로..


> [MoreFloorsTerrainsCarpet.zip](https://github.com/RimWorldKorea/RMK/files/13401880/MoreFloorsTerrainsCarpet.zip)
> 
> 해당 파일을 XMLister로 돌렸을 때 주석까지 전부 읽어들여버려요
> 
> <img alt="화면 캡처 2023-11-19 104315" width="1735" src="https://user-images.githubusercontent.com/150018627/284031290-8f28b622-a9df-4979-a83b-e95277074cfb.png">

- 이 두 개는 그러지 않도록 수정했습니다.

**Reply by csh1668** [reply]
@RimWorldKorea 
멘션하는걸 깜빡했네

**Reply by RimWorldKorea** [reply]
제가 따로 설정을 안 해서 그런지 멘션 안해도 알림은 전부 오더라구요 ㅎ
곧 대량으로 쓸 일이 생길거라 그 때 확인을..

**Comment by RimWorldKorea** [comment]
<img width="164" alt="화면 캡처 2023-11-22 025234" src="https://github.com/RimWorldKorea/RMK/assets/150018627/96c45f41-3d8a-49c0-bfd4-5761a5c9aa1f">
<img width="102" alt="화면 캡처 2023-11-22 025443" src="https://github.com/RimWorldKorea/RMK/assets/150018627/9d34e68c-4b81-4f8e-91b7-6315b334f254">

+ 세퍼레이터 설정에 관계없이 구분자가 ";"으로 고정되는 증상이 있어요!

+ 파일명에 특정 텍스트가 포함된 파일만을 선택적으로 병합하는 필터를 넣을 수 있을까요?
    예를 들어 폴더 안에 xml 파일이 여러개 있는데
    그 중에서 파일명에 TynanAss 라는 단어가 포함된 파일에 대해서만 작동하도록

**Reply by csh1668** [reply]
[net7.0.zip](https://github.com/RimWorldKorea/RMK/files/13438256/net7.0.zip)

수정사항을 업로드하는걸 깜빡했네요 ㅎ

**Reply by RimWorldKorea** [reply]
이정도면 대만족이네요!
작업이 23배는 빨라진 것 같아요

**Reply by csh1668** [reply]
또 뭐 프로그래밍적으로 필요한 거 있으시면 언제든 말씀해주세요 :)
